gcloud-java-bigquery/src/main/java/com/google/gcloud/spi/DefaultBigQueryRpc.java

```diff
@@ -0,0 +1,448 @@
+          .setFields(FIELDS.getString(options))
+          .setQuotaUser(QUOTA_USER.getString(options))
+          .setUserIp(USER_IP.getString(options))
+          .execute();
+    } catch(IOException ex) {
+      BigQueryException serviceException = translate(ex);
+      if (serviceException.code() == HTTP_NOT_FOUND) {
+        return null;
+      }
+      throw serviceException;
+    }
+  }
+
+  @Override
+  public Tuple<String, Iterable<Dataset>> listDatasets(Map<Option, ?> options)
+      throws BigQueryException {
+    try {
+      DatasetList datasetsList = bigquery.datasets()
+          .list(this.options.projectId())
+          .setQuotaUser(QUOTA_USER.getString(options))
+          .setUserIp(USER_IP.getString(options))
+          .setAll(Option.ALL_DATASETS.getBoolean(options))
+          .setMaxResults(MAX_RESULTS.getLong(options))
+          .setPageToken(PAGE_TOKEN.getString(options))
+          .execute();
+      Iterable<DatasetList.Datasets> datasets = datasetsList.getDatasets();
+      return Tuple.of(datasetsList.getNextPageToken(),
+          Iterables.transform(datasets != null ? datasets : ImmutableList.<DatasetList.Datasets>of(),
+              new Function<DatasetList.Datasets, Dataset>() {
+                @Override
+                public Dataset apply(DatasetList.Datasets f) {
+                  return new Dataset()
+                      .setDatasetReference(f.getDatasetReference())
+                      .setFriendlyName(f.getFriendlyName())
+                      .setId(f.getId())
+                      .setKind(f.getKind());
+                }
+              }));
+    } catch (IOException ex) {
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public Dataset create(Dataset dataset, Map<Option, ?> options) throws BigQueryException {
+    try {
+      return bigquery.datasets().insert(this.options.projectId(), dataset)
+          .setFields(FIELDS.getString(options))
+          .setQuotaUser(QUOTA_USER.getString(options))
+          .setUserIp(USER_IP.getString(options))
+          .execute();
+    } catch (IOException ex) {
+      throw translate(ex);
+    }
+  }
+
+  @Override
+  public boolean deleteDataset(String datasetId, Map<Option, ?> options) throws BigQueryException {
+    try {
+      bigquery.datasets().delete(this.options.projectId(), datasetId)
+          .setQuotaUser(QUOTA_USER.getString(options))
+          .setUserIp(USER_IP.getString(options))
+          .setDeleteContents(DELETE_CONTENTS.getBoolean(options))
+          .execute();
+      return true;
+    } catch (IOException ex) {
+      BigQueryException serviceException = translate(ex);
+      if (serviceException.code() == HTTP_NOT_FOUND) {
+        return false;
+      }
+      throw serviceException;
+    }
+  }
+
+  @Override
+  public Dataset patch(Dataset dataset, Map<Option, ?> options) throws BigQueryException {
+    try {
+      DatasetReference reference = dataset.getDatasetReference();
```

Comment:
Yes, same holds for Table.
I realized that in create methods I take the project's id from `BigQueryOption` while in patch I take it from `DatasetReference`, this might be quite an inconsistency so we should decide the way to go.

1. If we take the project id from `BigQueryOption` the request will fail if `dataset.getDatasetReference().getProjectId()` does not match
2. If we take the project id from `DatasetReference` we allow patching/creating for different projects. However, we do not support get for different projects as `getDataset(String datasetId)` takes the project id from `BigQueryOption`

I prefer option 1), do you have any thoughts?